File: bundle-deployer/src/validation/alert_rules_are_valid.ts
// Validate alert rules file structure

import type { Bundle } from "../types/bundle.ts";
import { getErrorMessage } from "../utils/error.ts";

interface AlertRulesFile {
  apiVersion: number;
  groups: Array<{
    orgId?: number;
    name: string;
    folder: string;
    interval: string;
    rules: Array<{
      uid: string;
      title: string;
      condition: string;
      data: unknown[];
      [key: string]: unknown;
    }>;
  }>;
}

export async function run(base: string, bundle: Bundle): Promise<void> {
  if (!bundle.alert_rules) {
    console.log("ℹ️  No alert rules defined in bundle (optional)");
    return;
  }
  
  const alertRulesPath = `${base}/${bundle.alert_rules.path}`;
  
  console.log(`Validating alert rules at ${alertRulesPath}...`);
  
  try {
    const content = await Deno.readTextFile(alertRulesPath);
    const alertRules = JSON.parse(content) as AlertRulesFile;
    
    // Validate structure
    if (!alertRules.apiVersion) {
      throw new Error("alert_rules.apiVersion is required");
    }
    
    if (!Array.isArray(alertRules.groups)) {
      throw new Error("alert_rules.groups must be an array");
    }
    
    if (alertRules.groups.length === 0) {
      throw new Error("alert_rules.groups must contain at least one group");
    }
    
    // Validate each group
    for (let i = 0; i < alertRules.groups.length; i++) {
      const group = alertRules.groups[i];
      
      if (!group.name) {
        throw new Error(`Group ${i}: name is required`);
      }
      
      if (!group.folder) {
        throw new Error(`Group ${i} (${group.name}): folder is required`);
      }
      
      if (!group.interval) {
        throw new Error(`Group ${i} (${group.name}): interval is required`);
      }
      
      if (!Array.isArray(group.rules)) {
        throw new Error(`Group ${i} (${group.name}): rules must be an array`);
      }
      
      if (group.rules.length === 0) {
        throw new Error(`Group ${i} (${group.name}): rules must contain at least one rule`);
      }
      
      // Validate each rule
      for (let j = 0; j < group.rules.length; j++) {
        const rule = group.rules[j];
        
        if (!rule.uid) {
          throw new Error(`Group ${group.name}, Rule ${j}: uid is required`);
        }
        
        if (!rule.title) {
          throw new Error(`Group ${group.name}, Rule ${j}: title is required`);
        }
        
        if (!rule.condition) {
          throw new Error(`Group ${group.name}, Rule ${j}: condition is required`);
        }
        
        if (!Array.isArray(rule.data)) {
          throw new Error(`Group ${group.name}, Rule ${j}: data must be an array`);
        }
      }
    }
    
    console.log(`✓ Alert rules file is valid (${alertRules.groups.length} group(s), ${getTotalRules(alertRules)} rule(s))`);
    
  } catch (e) {
    throw new Error(`Failed to validate alert rules at ${alertRulesPath}: ${getErrorMessage(e)}`);
  }
}

function getTotalRules(alertRules: AlertRulesFile): number {
  return alertRules.groups.reduce((sum, group) => sum + group.rules.length, 0);
}

File: src/validate/alert_rules_are_valid.rs
// Validate alert rules file structure

use serde::{Deserialize, Serialize};
use serde_json::Value;
use tokio::fs;

use crate::bundle_struct::Bundle;

#[derive(Debug, Serialize, Deserialize)]
struct AlertRulesFile {
    #[serde(rename = "apiVersion")]
    api_version: Option<i32>,
    groups: Vec<AlertGroup>,
}

#[derive(Debug, Serialize, Deserialize)]
struct AlertGroup {
    #[serde(rename = "orgId")]
    org_id: Option<i32>,
    name: String,
    folder: String,
    interval: String,
    rules: Vec<AlertRule>,
}

#[derive(Debug, Serialize, Deserialize)]
struct AlertRule {
    uid: String,
    title: String,
    condition: String,
    data: Vec<Value>,
}

pub async fn run(base: &str, bundle: &Bundle) -> Result<(), String> {
    if bundle.alert_rules.is_none() {
        println!("ℹ️  No alert rules defined in bundle (optional)");
        return Ok(());
    }

    let alert_rules_ref = bundle.alert_rules.as_ref().unwrap();
    let alert_rules_path = format!("{}/{}", base, alert_rules_ref.path);

    println!("Validating alert rules at {}...", alert_rules_path);

    let content = fs::read_to_string(&alert_rules_path)
        .await
        .map_err(|e| {
            format!(
                "Failed to read alert rules file at {}: {}",
                alert_rules_path, e
            )
        })?;

    let alert_rules: AlertRulesFile = serde_json::from_str(&content).map_err(|e| {
        format!(
            "Failed to parse alert rules JSON at {}: {}",
            alert_rules_path, e
        )
    })?;

    // Validate structure
    if alert_rules.api_version.is_none() {
        return Err("alert_rules.apiVersion is required".to_string());
    }

    if alert_rules.groups.is_empty() {
        return Err("alert_rules.groups must contain at least one group".to_string());
    }

    // Validate each group
    for (i, group) in alert_rules.groups.iter().enumerate() {
        if group.name.is_empty() {
            return Err(format!("Group {}: name is required", i));
        }

        if group.folder.is_empty() {
            return Err(format!("Group {} ({}): folder is required", i, group.name));
        }

        if group.interval.is_empty() {
            return Err(format!("Group {} ({}): interval is required", i, group.name));
        }

        if group.rules.is_empty() {
            return Err(format!(
                "Group {} ({}): rules must contain at least one rule",
                i, group.name
            ));
        }

        // Validate each rule
        for (j, rule) in group.rules.iter().enumerate() {
            if rule.uid.is_empty() {
                return Err(format!("Group {}, Rule {}: uid is required", group.name, j));
            }

            if rule.title.is_empty() {
                return Err(format!(
                    "Group {}, Rule {}: title is required",
                    group.name, j
                ));
            }

            if rule.condition.is_empty() {
                return Err(format!(
                    "Group {}, Rule {}: condition is required",
                    group.name, j
                ));
            }
        }
    }

    let total_rules = get_total_rules(&alert_rules);
    println!(
        "✓ Alert rules file is valid ({} group(s), {} rule(s))",
        alert_rules.groups.len(),
        total_rules
    );

    Ok(())
}

fn get_total_rules(alert_rules: &AlertRulesFile) -> usize {
    alert_rules
        .groups
        .iter()
        .map(|group| group.rules.len())
        .sum()
}
